commit 20b77570320a11681209765f37bb893d22c3ed56
Author: Dag Wieers <dag@wieers.com>
Date:   Tue Mar 8 00:28:41 2016 +0100

    Ensure that a download failure is properly raised before the read fails
    
    Without this change, a download failure may bail out with the message:
    
        "Failure downloading http://foo/bar, 'NoneType' object has no attribute 'read'"
    
    whereas with this fix, you'd get a proper error like:
    
        "Failure downloading http://foo/bar, Request failed: <urlopen error [Errno 113] No route to host>"
    
    or one of the many other possible download errors that can occur.

diff --git a/lib/ansible/modules/files/unarchive.py b/lib/ansible/modules/files/unarchive.py
index 251fbab35e..e4bb230aef 100644
--- a/lib/ansible/modules/files/unarchive.py
+++ b/lib/ansible/modules/files/unarchive.py
@@ -282,6 +282,9 @@ def main():
             package = os.path.join(tempdir, str(src.rsplit('/', 1)[1]))
             try:
                 rsp, info = fetch_url(module, src)
+                # If download fails, raise a proper exception
+                if rsp is None:
+                    raise Exception(info['msg'])
                 f = open(package, 'w')
                 # Read 1kb at a time to save on ram
                 while True:
